src/lib.rs

```diff
@@ -19,288 +23,241 @@ use syntax::ast_map;
-                        let attrs = get_attrs_opt(self.cx.tcx, id);
-                        if let Some(v) = attrs {
-                            if v.iter().any(|item| item.check_name("allowed_on_protected")) {
-                                for param in params {
-                                    if let ExprPath(_, _) = param.node {
-                                        // It's an ident within an allowed method call,
-                                        // it's fine!
-                                        self.cx.tcx.sess.span_note(ex.span, "Allowed usage of type. Carry on!");
-                                        return;
-                                    } else {
-                                        // It could be a situation like
-                                        // allowed_fn(foo, bar, {foo(); bar(); unsafe_drop_fn(protected); baz()})
-                                        // this ensures that no trickery happens
-                                        self.visit_expr(&*param);
-                                    }
-                                }
-                            } else if v.iter().any(|item| item.check_name("allowed_drop")) {
-                                for param in params {
-                                    if let ExprPath(_, _) = param.node {
-                                        // It's an ident within an allowed drop call,
-                                        // we should remove it from the map if it was there
-                                        let def = self.cx.tcx.def_map.borrow().get(&param.id).map(|&x| x);
-                                        if let Some(PathResolution { base_def: DefLocal(id), .. }) = def {
-                                            self.cx.tcx.sess.span_note(ex.span, "Properly dropped!");
-                                            self.map.remove(&id);
-                                            return;
-                                        }
-                                    } else {
-                                        self.visit_expr(&*param)
-                                    }
-                                }
-                            }
-                        }
+            ExprCall(_, _) | ExprMethodCall(_, _, _) => {
+                visit::walk_expr(self, e);
+            }
+            ExprAddrOf(_, _) => {
+                // Ignore children
+                return;
+            }
+            ExprIf(ref e1, ref b, ref else_block) => {
+                // Consume stuff in expr
+                visit::walk_expr(self, &e1);
+
+                // visit block(s)
+                if let &Some(ref e2) = else_block {
+                    // The resulting hms should be the same
+                    let mut v1 = self.clone();
+                    let mut v2 = self.clone();
+                    v1.visit_block(&b);
+                    v2.visit_expr(&e2);
+                    if v1.map != v2.map {
```

Comment:
So simple; I should have thought of that! :smile: 

Might lead to lots of churn but meh, compile time perf is okay.

Also, be careful on how equality on hashmaps is calculated, it might not be what we expect.